### ISSUE TITLE
Separate indent options for var, let and const

### ISSUE THREAD

**Issue by scottohara** [eslint/eslint#3339]
In my variable declarations, I prefer to vertically align the identifiers within each `var`, `let` or `const` declaration, eg.
```
// Note: '.' below represents a tab character

var foo = 1,  // <- 'foo' aligns with 'bar'
. . bar = 2;

let baz = 3,  // <- 'baz' aligns with 'qux'
. . qux = 4;

const a = 5,  // <- 'a' aligns with 'b'
. . . b = 6;
```
The following rule in ESLint@1.1.0:
```
"indent": [2, "tab", {"VariableDeclarator": 2}]
```
...yields the error `Expected indentation of 2 characters but found 3. (indent)`.
This is because the const `b` is indented by 3 tabs instead of 2 to maintain alignment with `a` (due to the keyword `const` being longer than both `var` and `let`).

(Note: I prefer tabs over spaces, but the same issue would apply to spaces also)

Just wondering if anyone else shares this preference (for lining up identifiers vertically), and if so, whether consideration should be given to expanding the `VariableDeclarator` option to allow individually specified indents for `var`, `let` and `const`, eg.
```
"indent": [2, "tab", {"VariableDeclarator": {"var": 2, "let": 2, "const": 3}}]
```
If not, alternatives that don't violate the `indent` rule would be either:
```
// VariableDeclarator: 3, with extra indentation on top lines of any var/let declarations 

/* eslint indent: [2, "tab", {"VariableDeclarator": 3}]*/

var. .foo = 1,  // <- Extra spacing between 'var' and 'foo'
. . . bar = 2;

let. .baz = 3,  // <- Extra spacing between 'let' and 'baz'
. . . qux = 4;

const a = 5,
. . . b = 6;
```
```
// VariableDeclarator: 2, and ignore const top line alignment, 

/* eslint indent: [2, "tab", {"VariableDeclarator": 2}]*/

var foo = 1,
. . bar = 2;

let baz = 3,
. . qux = 4;

const a = 5,  // <- 'a' doesn't align with 'b'...
. . b = 6;    // <- ...but 'b' does align with 'foo', 'bar', 'baz', etc.
```

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
This was a known limitation when we rewrote the rule. We still haven't quite figured out how to make these separately configurable.

**Comment by BYK** [comment]
@nzakas - if you're okay with the configuration schema I have a working prototype for this.

**Comment by IanVS** [comment]
Does it really make sense to separately configure each of `let`, `var`, and `const`?  It sure would be nice if there could be a single option which would account for all of them.  Something like `"AlignVariableDeclarators": true`.  Specifying numbers for all of these seems very clunky...

**Comment by gyandeeps** [comment]
Single option is already available.

**Comment by IanVS** [comment]
Yes, that's what I mean.  It would be ideal to have one option which handles all three properly.  Is that impossible?

**Comment by gyandeeps** [comment]
Everything is possible but its just having 3 opens up for more option combinations. I dont have preference either way as I prefer separate declaration for every variable.

**Comment by BYK** [comment]
In all honesty I don't think this new addition would be useful except for the following configuration: `[2, { "VariableDeclarator": { "var": 2, "let": 2, "const": 3 }`. This is mostly because the original behavior was implemented as a multiplier to the original indent (which I kept in #3394 for bw. compatibility).

I think either automatically detecting the "preferred" indent and warning on that based on an option would be much useful as @IanVS suggested. `undefined` would indicate no preference where as `false` would forbid and `true` would enforce it.

Closing #3394 (without merging), deprecating the original `VariableDeclarator` option and then implementing this idea seems the better way forward.

**Comment by BYK** [comment]
@nzakas - no comments regarding https://github.com/eslint/eslint/issues/3339#issuecomment-131549701? :)

**Comment by dionyziz** [comment]
This seems to work with the 2-space indentation because `VariableDeclarator` sets a multiple of the indentation level. However, we use 4 spaces for indentation and we still align our declaration statements:

```
    const a = 1,
          b = 2;
```

The configuration suggested hence does not work and there's no configuration that works, as 6 is not a multiple of 4.

**Comment by BYK** [comment]
Totally agree to @dionyziz. @eslint/eslint-team - I think we should schedule a breaking change for this to make these options absolute values instead of multipliers to the current indentation.

**Comment by mysticatea** [comment]
I think, this can be fixed modifying schema, from `integer` to `number`.

```js
module.exports.schema = [
    {
        "oneOf": [
            {
                "enum": ["tab"]
            },
            {
                "type": "integer"
            }
        ]
    },
    {
        "type": "object",
        "properties": {
            "SwitchCase": {
                "type": "number"
            },
            "VariableDeclarator": {
                "type": ["number", "object"],
                "properties": {
                    "var": {
                        "type": "number"
                    },
                    "let": {
                        "type": "number"
                    },
                    "const": {
                        "type": "number"
                    }
                }
            }
        },
        "additionalProperties": false
    }
];
```

```js
/* eslint-env es6 */
/* eslint indent: [2, 4, {VariableDeclarator: 1.5}] */
const a = 1,
      b = 2;
```

```
> node bin/eslint.js test.js --no-eslintrc
>
```

no errors.

**Comment by BYK** [comment]
@mysticatea - using fractions fix the issue yeah, but the logic doesn't really make sense to me so I still prefer using absolute numbers.

**Comment by dionyziz** [comment]
Can we reopen this issue if this is a confirmed problem?

**Comment by DaVarga** [comment]
+1 reopen

**Comment by soluml** [comment]
+1 from me as well.  It'd be nice if VariableDeclarator could just be separate from the indent rules (instead of a multiplier), but in the interim changing VariableDeclarator to allow numbers (then rounding to nearest whole number) would be ideal:

    "indent": [
        2,
        4,
        {
            "SwitchCase": 1,
            "VariableDeclarator": {"var": 1, "let": 1, "const": 1.5}
        }
    ],

**Comment by oh-ren** [comment]
Agreed. But as the original issue has been resolved, perhaps a new issue should be created for this?

If `indent` *will* undergo breaking changes, please also consider #4496

**Comment by fastfedora** [comment]
+1 for allowing decimal values in the interim. I'm also stuck in the situation of using 4 spaces for indentation, but want to align with the declaration. Just tested in v3.9.1 and using `1.5` doesn't seem to work.

**Comment by Ohar** [comment]
Reopen please. 
Float koefs fix it ofc.

**Comment by pcellis** [comment]
+1 reopen and for the decimal value fix in the schema. It's incredibly annoying to have to start each `const` variable declaration with the `const` keyword in order to get eslint to shut up about the indent issue.

**Comment by platinumazure** [comment]
@pcellis Please open a new issue, thanks!

**Comment by Ohar** [comment]
@platinumazure Why new issue? The problem still the same.

**Comment by platinumazure** [comment]
Closed issues get lost in the backlog, and issues with a lot of old
discussion already can be confusing to work with. Instead, we welcome the
creation of new issues if someone wants to restart a discussion. Thanks!
